design/build-cache.md

```diff
@@ -0,0 +1,201 @@
+ss# Build cache
+
+## Context
+
+Users require a way to cache resources between jobs within a single pipeline or consecutive executions of a particular pipeline in order to speed up build time.
+
+## Status
+
+- 29 June 2018: Proposal submitted
+
+## Proposal
+
+Provide a mechanism for transparent caching of resources between jobs within a single pipeline, or between consecutive executions of a particular pipeline.
+
+1. Content to be cached can be specified by the user.
+2. Caching is scoped. 
+   1. Job level caching
+   2. Event level caching
+   3. Pipeline level caching (static caching, in a sense)
+3. Well used caching patterns can be offered as predefined options.
+   1. Dependency caching for build systems such as npm, sbt and cargo.
+
+The build cache and cache usage is configurable via Yaml and backed by an internal sd-command and caching service.
+
+## Details
+
+### Yaml specification
+
+```yml
+shared:
+  # Caches defined in the shared section can be pipeline or exection scoped.
+  cache:
+    # A pipeline scoped cache.
+    pipeline:  
+      # The caching targets
+      target: ["build/**", "~/.sbt"]
+    # An event scoped cache
+    event:
+      target: ["target/deployment.jar"]
+    # A job scoped cache 
+    my-cache-1:
+      target: ["build/some-artifact.zip"]
+jobs:
+  test:
+    requires: [~pr]
+    image: rust:latest
+    steps:
+      - test: cargo test
+  build: 
+    requires: [~pr]
+    # Indicate which job scoped caches to use.
+    cache: ["my-cache-1"]
+    image: rust:latest
+    steps:
+      - test: cargo build
+  publish-preview:
+    # Indicate which job scoped caches to use.
+    cache: ["my-cache-1"]
+    requires: [~commit]
```

Comment:
I think the cache does not work in this example. `build` is triggered by a pull request event and `publish-preview` is triggered by a commit event. According to scoping section, job level scoping works within a single event. These events are difference so SD creates caches in each job and these caches are not shared.

Reply:
Yeah, this example is meant to be a syntax proposal. Consequently, I am leaving out addition jobs that might use caches and just serve to increase the amount of intellectual load required to parse the example.

Check the comment in line 107. I'm proposing a slightly different way of scoping the caches.